lib/rules/selector-attribute-quotes/index.js

```diff
@@ -64,6 +65,17 @@ const rule = (primary, _secondaryOptions, context) => {
 					}
 
 					if (quoted && primary === 'never') {
+						// some selectors require quotes to be valid;
+						// we pass in the raw string value, which contains the escape characters
+						// necessary to check if escaped characters are valid
+						// see: https://github.com/stylelint/stylelint/issues/4300
+						if (
+							!attributeNode.raws.value ||
+							!isValidIdentifier(attributeNode.raws.value.slice(1, -1))
+						) {
```

Comment:
Admittedly, this feels off - but I (think?) that the `raws` is the only way I can keep the escape characters in the value (opposed to `attributeNode.value`, which unescapes everything) so I can check if the resulting string is a valid identifier, etc. Suggestions on how to make this clearer are welcome!

Reply:
[suggest] What about using the [`Attribute#smartQuoteMark()` method of `postcss-selector-parser`](https://github.com/postcss/postcss-selector-parser/blob/v6.0.11/src/selectors/attribute.js#L132) instead of the custom logic?

```js
// The value ident is invalid if the attribute needs a quote mark.
if (typeof attributeNode.smartQuoteMark({}) === 'string') {
	return;
}
```

The method doc says:

> Intelligently select a quoteMark value based on the value's contents. If the value is a legal CSS ident, it will not be quoted. Otherwise a quote mark will be picked that minimizes the number of escapes.

The method's implementation relies on the [`cssesc`](https://github.com/mathiasbynens/cssesc) package, which seems reliable and popular (22M downloads/week).

Reply:
Thanks for the suggestion @ybiquitous, this is very cool - I hadn't heard about it before!

However, I'm concerned that it doesn't completely resolve our problem. In particular, if I use that instead of `isValidIdentifier`, the [cases identified by \@Mouvedia](https://github.com/stylelint/stylelint/issues/4300#issuecomment-1379256755) are not handled well:

```js
[href='te\\'s\\'t'] { }  // currently autofixable, now missed
[href="te\\"s\\"t"] { }  // currently autofixable, now missed
```

I think this is because the [`Attribute#smartQuoteMark()` method](https://github.com/postcss/postcss-selector-parser/blob/v6.0.11/src/selectors/attribute.js#L132) uses the `value` of the attribute, and not the `raws`; the escaped quotes aren't preserved, and so it incorrectly believes that the attribute is not autofixable.

However, I may be using the function wrong or am missing something. Do you have any suggestions?

Reply:
If cssesc doesn't do the job just keep what you had, it's probably good enough.

IMHO adding a new dependency should have a minimum requirement.
e.g. used by 3 rules